pkgdb/src/fetchers/wrapped-nixpkgs-input.cc

```diff
@@ -0,0 +1,543 @@
+ *
+ *
+ * -------------------------------------------------------------------------- */
+
+#include <algorithm>
+#include <cctype>
+#include <cstdint>
+#include <filesystem>
+#include <fstream>
+#include <map>
+#include <memory>
+#include <optional>
+#include <regex>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include <nix/attrs.hh>
+#include <nix/cache.hh>
+#include <nix/error.hh>
+#include <nix/fetchers.hh>
+#include <nix/flake/flake.hh>
+#include <nix/flake/flakeref.hh>
+#include <nix/flake/lockfile.hh>
+#include <nix/hash.hh>
+#include <nix/logging.hh>
+#include <nix/path.hh>
+#include <nix/ref.hh>
+#include <nix/store-api.hh>
+#include <nix/types.hh>
+#include <nix/url-parts.hh>
+#include <nix/url.hh>
+#include <nix/util.hh>
+
+#include "flox/core/nix-state.hh"
+#include "flox/core/util.hh"
+
+
+/* -------------------------------------------------------------------------- */
+
+/* Forward declaration */
+namespace nix {
+class EvalState;
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+namespace flox {
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Create a temporary directory containing a `flake.nix` which wraps
+ *        @a nixpkgsRef configuring it to allow unfree and broken packages.
+ */
+static std::filesystem::path
+createWrappedFlakeDirV0( const nix::FlakeRef & nixpkgsRef )
+{
+  /* Create a temporary directory to put the filled out template file in it. */
+  std::filesystem::path tmpDir = nix::createTempDir();
+  debugLog( "created temp dir for flake template: " + tmpDir.string() );
+
+  /* Fill out the template with the flake references. */
+  std::ofstream            flakeOut( tmpDir / "flake.nix" );
+  static const std::string flakeTemplate =
+#include "./flake-v0.nix.in.hh"
+    ;
+  std::istringstream flakeIn( flakeTemplate );
+  std::string        line;
+  while ( std::getline( flakeIn, line ) )
+    {
+      /* Inject URL */
+      if ( line.find( "@NIXPKGS_URL@" ) != std::string::npos )
+        {
+          line.replace( line.find( "@NIXPKGS_URL@" ),
+                        std::string( "@NIXPKGS_URL@" ).length(),
+                        nixpkgsRef.to_string() );
+        }
+      flakeOut << line << '\n';
+    }
+  flakeOut.close();
+  debugLog( "filled out flake template with flake reference: "
+            + nixpkgsRef.to_string() );
+
+  /* Lock the filled out template to avoid spurious re-locking and silence the
+   * "Added input ..." message. */
+  flox::NixState           nixState;
+  nix::ref<nix::EvalState> state = nixState.getState();
+  nix::FlakeRef wrappedRef = nix::parseFlakeRef( "path:" + tmpDir.string() );
+  /* Push verbosity level to suppress "warning: creating lock file ..." */
+  auto oldVerbosity = nix::verbosity;
+  nix::verbosity    = nix::lvlError;
+  auto _locked      = nix::flake::lockFlake( *state, wrappedRef, {} );
```

Comment:
Should we be using the new wrapper `flox::lockFlake` to ensure downloads (if any) occur in a child process?

Reply:
Yes

Reply:
Fixed